spec/index.adoc

```diff
@@ -4199,18 +4199,52 @@ Read response::
 * If `F(Q.Arg, Q.sender, Env) = Trap trap` then
 +
 ....
-{status: rejected; reject_code: CANISTER_ERROR, reject_message: <implementation-specific>, error_code: <implementation-specific>, node_signatures: <implementation-specific>}
+{status: "rejected"; reject_code: CANISTER_ERROR, reject_message: <implementation-specific>, error_code: <implementation-specific>, node_signatures: <implementation-specific>}
 ....
 * Else if `F(Q.Arg, Q.sender, Env) = Return {response = Reject (code, msg); …}` then
 +
 ....
-{status: rejected; reject_code: <code>: reject_message: <msg>, error_code: <implementation-specific>, node_signatures: <implementation-specific>}
+{status: "rejected"; reject_code: <code>: reject_message: <msg>, error_code: <implementation-specific>, node_signatures: <implementation-specific>}
 ....
 * Else if `F(Q.Arg, Q.sender, Env) = Return {response = Reply R; …}` then
 +
 ....
-{status: replied; reply: { arg :  <R> }, node_signatures: <implementation-specific>}
-....
+{status: "replied"; reply: {arg: <R>}, node_signatures: <implementation-specific>}
+....
+
+Given a certificate obtained by requesting the path `/subnet` via a read state request
+to `/api/v2/canister/<ECID>/read_state`, the following predicate describes
+when the returned response is correctly signed by the nodes:
+....
+verify_response({status: "replied"; reply: {arg: <R>}, node_signatures: Sigs}, Cert)
+  = verify_cert(Cert) ∧
+    ((Cert.delegation = NoDelegation ∧ SubnetId = RootSubnetId ∧ lookup(["subnet",SubnetId,"canister_ranges"], Cert) = Found Ranges) ∨
+     (SubnetId = Cert.delegation.subnet_id ∧ lookup(["subnet",SubnetId,"canister_ranges"], Cert.delegation.certificate) = Found Ranges)) ∧
```

Comment:
My tendency would be to also mandate including a self delegation for the NNS to avoid the special case here.

Reply:
Would you then also mandate an NNS self delegation for certificates passed to canisters in query calls (carrying the canister's certified variables)?

Reply:
Valid point, but given that this part is like the existing certification procedure, I'd not change it here, so that the same verification code can be used

Reply:
> Valid point, but given that this part is like the existing certification procedure, I'd not change it here, so that the same verification code can be used

Wouldn't the existing verification procedure just transparently work with a self-delegation of the NNS?

> Would you then also mandate an NNS self delegation for certificates passed to canisters in query calls (carrying the canister's certified variables)?

Assuming my point above is correct, we could only require the NNS delegation to be present in this verification algorithm here, but not elsewhere. In summary, I think that if we can make this somehow elegantly work without making the scope of this change too broad it would be nice to have. Otherwise, I'm also happy with this special case here.

Reply:
When you say NNS subnet, you mean root subnet, right? (It happens that the NNS canisters live on the root subnet, but that's more a coincidence than a necessary and could change in the future, and here the root-of-trust aspect is the relevant one.)

I'm still not convinced. The gain is very small; the code already has to be able to check certificates with delegation and certificates without (at least to check the delegations). Forcing the delegation chain to have length 1 doesn't save much here, but incurs relatively pointless additional work and bandwidth (in the form of the self-delegation).

Reply:
> When you say NNS subnet, you mean root subnet, right?

Yes. Sorry. 

> I'm still not convinced. [...]

As said above, I don't have a very strong opinion on this. Lets keep what we have now.